Improve quest drop notification

SabreCat opened the pull request:
1. Changes the notification of quests awarded on levelup to use a modal instead of a drop-style alert.
2. Fixes (I think!) the bug where quest scroll drops appeared twice, then reset to 1 in inventory upon refresh.
3. Refactors the quests Jade file to use a mixin for reused quest information readout.
4. ~~Refactors drop notifications to use `switch` instead of a bunch of `else if`s.~~

SabreCat committed:
feat(quests): Improve drop notif

SabreCat committed:
fix(quests): Remove controller instantiation

SabreCat commented on website/views/shared/modals/quests.jade:
```diff
@@ -90,3 +77,17 @@ script(type='text/ng-template', id='modals/questInvitation.html')
+    p=env.t('dropQuestCongrats')
+  .modal-footer
+    button.btn.btn-default(ng-click='closeQuest(); $close()')=env.t('questLater')
+    button.btn.btn-primary(ng-click='questInit(); $close()')=env.t('inviteParty')
```
I wanted to have this conditionally be the 'create a party' button, but `inviteOrStartParty` wasn't working here.

crookedneighbor replied:
I don't think you're passing in a controller when opening this modal. Would it be helpful to pass in the partyCtrl and create a function on `$scope` there that does the `inviteOrStartParty` code? This is how it's done in the Header controller: https://github.com/HabitRPG/habitrpg/blob/32a74471dbeb791fe5fafc6d5441e5cda827968b/website/public/js/controllers/headerCtrl.js#L9

SabreCat replied:
I can give that a try!

gisikw commented on website/public/js/controllers/notificationCtrl.js:
```diff
@@ -58,9 +58,11 @@ habitrpg.controller('NotificationCtrl',
-      if (after == before || !after) return;
-      $rootScope.playSound('Achievement_Unlocked');
-      if (after.type !== 'gear') {
+      if (_.isEqual(after, before) || !after) return;
```
Why not `===`?

SabreCat replied:
It wasn't working, in mysterious fashion. I would see the `if` statement blow right by, despite both `after` and `before` being, e.g. {type:'Quest',key:'atom1'}. Changing it to `_.isEqual` stopped that happening, eliminating the double-notification bug.

gisikw commented on website/public/js/controllers/notificationCtrl.js:
```diff
@@ -70,23 +72,30 @@ habitrpg.controller('NotificationCtrl',
-      }else{
-        // Keep support for another type of drops that might be added
-        Notification.drop(User.user._tmp.drop.dialog);
+      switch(after.type) {
```
Booo switch statements (honestly, I'd prefer we avoid switch statements. It's far too easy to forget a break and failthrough. Especially with OSS).

SabreCat replied:
Ah, if that's a bad pattern, I can revert it. I find it more readable than an `else if` chain, but I'm just a scrub, heh.

crookedneighbor replied:
I agree that it's easier to parse, but ultimately a bad pattern.

gisikw replied:
One alternative is the hashdict lookup:

``` js
handling = {
  Quest: function() {
    $rootScope.selectedQuest = Content.quests[after.key];
    $rootScope.openModal('questDrop');
  }
  // etc
};

handling[after.type]();
```

But that would probably require a bit of refactoring.

crookedneighbor commented on website/public/js/controllers/notificationCtrl.js:
```diff
@@ -58,9 +58,11 @@ habitrpg.controller('NotificationCtrl',
+      $rootScope.playSound('Item_Drop');
+
+      if (after.type !== 'gear' && after.type !== 'Quest') {
         var type = (after.type == 'Food') ? 'food' :
```
IMO, once ternaries turn into double ternaries, it's time to go with something that is not a ternary.

SabreCat replied:
OK, I'll refactor that into another `if`/`else if` chain.

crookedneighbor commented on website/views/shared/modals/quests.jade:
```diff
@@ -1,5 +1,23 @@
 include ./quest-rewards
 
+mixin questInfo
```
so pleased about this. :heart:

SabreCat committed:
fix(quests): Refactoring

Based on feedback from @crookedneighbor and @gisikw. Removes double ternary in the block that normalizes the item type string, and reverts the notification logic to if/else if format instead of switch/case. Also removes a block of commented-out code once used for death notifications.

crookedneighbor commented on website/public/js/controllers/notificationCtrl.js:
```diff
@@ -58,35 +58,42 @@ habitrpg.controller('NotificationCtrl',
+          type = 'food';
+        } else if (after.type === 'HatchingPotion') {
+          type = 'hatchingPotions';
+        } else type = after.type.toLowerCase() + 's';
```
Doesn't food fall into this?

crookedneighbor replied:
Oh never mind. Foods

crookedneighbor replied:
Id prefer that the last else had brackets.

crookedneighbor commented on website/public/js/controllers/notificationCtrl.js:
```diff
@@ -58,35 +58,42 @@ habitrpg.controller('NotificationCtrl',
         User.user.items[type][after.key]++;
       }
 
-      if(after.type === 'HatchingPotion'){
```
I think we can do some meta programming here to mitigate this awful if/else block, but that can be a refactor we do after this is merged in.

crookedneighbor commented on website/public/js/controllers/notificationCtrl.js:
```diff
@@ -121,17 +128,6 @@ habitrpg.controller('NotificationCtrl',
       $rootScope.openModal('achievements/contributor');
     });
 
-    /*_.each(['weapon', 'head', 'chest', 'shield'], function(watched){
```
Death to commented out code!

SabreCat replied:
You're like a one-man angry mob, calling for the death of various things that displease you!

gisikw replied:
Death to commented out code!

gisikw replied:
(commented code is a artifact from pre-version-control days)

SabreCat committed:
feat(quests): Invite party button

SabreCat commented on website/public/js/services/questServices.js:
```diff
@@ -84,12 +85,22 @@
       showQuest: showQuest,
-      closeQuest: closeQuest
+      closeQuest: closeQuest,
+      questInit: questInit
```
Implementing the Invite Party button on the quest drop modal involved moving another function out of the controllers into the Quests service. How does this look?

SabreCat commented on website/public/js/services/questServices.js:
```diff
@@ -84,12 +85,22 @@
+      $rootScope.party.$questAccept({key:$rootScope.selectedQuest.key}, function(){
+        $rootScope.party.$get();
+      });
+      $rootScope.$state.go('options.social.party');
```
Inviting the party to your quest had a curious lack of user feedback; I added this so we jump to the Party page. Does that look right?

SabreCat committed:
fix(quest): Move party jump to cb
